langchain/src/callbacks/index.ts

```diff
@@ -0,0 +1,259 @@
+import { AgentAction, AgentFinish } from "../agents/index.js";
+import { ChainValues } from "../chains/index.js";
+import { LLMResult } from "../schema/index.js";
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type Error = any;
+
+export interface BaseCallbackHandlerInput {
+  alwaysVerbose?: boolean;
+  ignoreLLM?: boolean;
+  ignoreChain?: boolean;
+  ignoreAgent?: boolean;
+}
+
+export abstract class BaseCallbackHandler implements BaseCallbackHandlerInput {
+  alwaysVerbose = false;
+
+  ignoreLLM = false;
+
+  ignoreChain = false;
+
+  ignoreAgent = false;
+
+  constructor(input?: BaseCallbackHandlerInput) {
+    if (input) {
+      this.alwaysVerbose = input.alwaysVerbose ?? this.alwaysVerbose;
+      this.ignoreLLM = input.ignoreLLM ?? this.ignoreLLM;
+      this.ignoreChain = input.ignoreChain ?? this.ignoreChain;
+      this.ignoreAgent = input.ignoreAgent ?? this.ignoreAgent;
+    }
+  }
+
+  handleLLMStart?(
+    llm: { name: string },
+    prompts: string[],
+    verbose?: boolean
+  ): Promise<void>;
+
+  handleLLMNewToken?(token: string, verbose?: boolean): Promise<void>;
+
+  handleLLMError?(err: Error, verbose?: boolean): Promise<void>;
+
+  handleLLMEnd?(output: LLMResult, verbose?: boolean): Promise<void>;
+
+  handleChainStart?(
+    chain: { name: string },
+    inputs: ChainValues,
+    verbose?: boolean
+  ): Promise<void>;
+
+  handleChainError?(err: Error, verbose?: boolean): Promise<void>;
+
+  handleChainEnd?(outputs: ChainValues, verbose?: boolean): Promise<void>;
+
+  handleToolStart?(
+    tool: { name: string },
+    input: string,
+    verbose?: boolean
+  ): Promise<void>;
+
+  handleToolError?(err: Error, verbose?: boolean): Promise<void>;
+
+  handleToolEnd?(output: string, verbose?: boolean): Promise<void>;
+
+  handleText?(text: string, verbose?: boolean): Promise<void>;
+
+  handleAgentAction?(action: AgentAction, verbose?: boolean): Promise<void>;
+
+  handleAgentEnd?(action: AgentFinish, verbose?: boolean): Promise<void>;
+}
+
+export abstract class BaseCallbackManager extends BaseCallbackHandler {
+  abstract addHandler(handler: BaseCallbackHandler): void;
+
+  abstract removeHandler(handler: BaseCallbackHandler): void;
+
+  abstract setHandlers(handlers: BaseCallbackHandler[]): void;
+
+  setHandler(handler: BaseCallbackHandler): void {
+    return this.setHandlers([handler]);
+  }
+}
+
+export class CallbackManager extends BaseCallbackManager {
+  handlers: BaseCallbackHandler[];
+
+  constructor() {
+    super();
+    this.handlers = [];
+  }
+
+  async handleLLMStart(
+    llm: { name: string },
+    prompts: string[],
+    verbose?: boolean
+  ): Promise<void> {
+    for (const handler of this.handlers) {
+      if (!handler.ignoreLLM && (verbose || handler.alwaysVerbose)) {
+        await handler.handleLLMStart?.(llm, prompts);
```

Comment:
How about adding a try catch around each call to a handler. That way of one handler errors your chain run (or whatever it was) doesn't fail, and a failure in one handler doesn't prevent calling the next handlers in the list

Reply:
Great idea

Reply:
resolved in followup PR